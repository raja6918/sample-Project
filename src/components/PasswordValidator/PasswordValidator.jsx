import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import CheckIcon from '@material-ui/icons/Check';

const Container = styled.div`
  padding-top: 10px;
  padding-bottom: 5px;
`;

const Row = styled.div`
  display: flex;
  align-items: center;
  font-size: 12px;
  font-weight: 500;
  color: #000005;
  letter-spacing: 0;
  line-height: 14px;
`;

const DotIcon = styled.span`
  padding-right: 15px;
  padding-left: 5px;
`;

/**
 * This ia a generic Password Strength Validator component to check whether user input satisfies set of hard-coded criteria.
 */
const PasswordValidator = props => {
  // labels and state boolean corresponding to each validation
  const validationCriteria = [
    [1, props.t('PASSWORD_VALIDATOR.rules.contains8Chars'), 'contains8Chars'],
    [
      2,
      props.t('PASSWORD_VALIDATOR.rules.containsLowerUpperCase'),
      'containsLowerUpperCase',
    ],
    [3, props.t('PASSWORD_VALIDATOR.rules.containsNum'), 'containsNum'],
    [4, props.t('PASSWORD_VALIDATOR.rules.containsNoBlank'), 'containsNoBlank'],
  ];

  const validatePassword = data => {
    let containsUpperCase = false;
    let containsLowerCase = false;
    let contains8Chars = false;
    let containsNum = false;
    let containsNoBlank = false;

    // has uppercase letter
    if (data.toLowerCase() !== data) {
      containsUpperCase = true;
    }

    // has lowercase letter
    if (data.toUpperCase() !== data) {
      containsLowerCase = true;
    }

    // has number
    if (/\d/.test(data)) {
      containsNum = true;
    }

    // has 8 characters
    if (data.length >= 8) {
      contains8Chars = true;
    }

    // Contain no blank spaces
    if (/^\S{1,}$/.test(data)) {
      containsNoBlank = true;
    }

    return {
      containsLowerUpperCase: containsLowerCase && containsUpperCase,
      contains8Chars,
      containsNum,
      containsNoBlank,
    };
  };

  const validityObj = validatePassword(props.value);

  return (
    <Container>
      {validationCriteria.map(validation => {
        const [id, title, isValid] = validation;
        const isValidCriteria = validityObj[isValid];
        const style = {
          color: isValidCriteria ? '#68c163' : '#00000061',
        };
        return (
          <Row key={id} style={style}>
            {isValidCriteria && (
              <CheckIcon
                color="inherit"
                fontSize="small"
                style={{ width: '25px', paddingRight: '10px' }}
              />
            )}
            {!isValidCriteria && <DotIcon>â€¢</DotIcon>}
            {title}
          </Row>
        );
      })}
    </Container>
  );
};

PasswordValidator.propTypes = {
  t: PropTypes.func.isRequired,
  value: PropTypes.string.isRequired,
};

export default PasswordValidator;
