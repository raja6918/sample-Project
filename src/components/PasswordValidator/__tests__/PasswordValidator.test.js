import React from 'react';
import { shallow } from 'enzyme';
import PasswordValidator from '../PasswordValidator';

describe('Test PasswordValidator Component', () => {
  const minProps = {
    t: jest.fn(msg => msg),
  };

  test('PasswordValidator Component renders correctly', () => {
    const wrapper = shallow(
      <PasswordValidator value="Test@12" {...minProps} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  test('Check whether PasswordValidator Component validates 8 characters when 8 letters provided', () => {
    const wrapper = shallow(
      <PasswordValidator value="12345678" {...minProps} />
    );
    // Title should be green
    expect(
      wrapper
        .find('PasswordValidator__Row')
        .at(0)
        .prop('style')
    ).toEqual({
      color: '#68c163',
    });
    expect(
      wrapper
        .find('PasswordValidator__Row')
        .at(0)
        .html()
    ).toContain('PASSWORD_VALIDATOR.rules.contains8Chars');
  });

  test('Check whether PasswordValidator Component validates 8 characters when 8 letters are not provided', () => {
    const wrapper = shallow(
      <PasswordValidator value="1234567" {...minProps} />
    );
    // Title should be grey
    expect(
      wrapper
        .find('PasswordValidator__Row')
        .at(0)
        .prop('style')
    ).toEqual({
      color: '#00000061',
    });
    expect(
      wrapper
        .find('PasswordValidator__Row')
        .at(0)
        .html()
    ).toContain('PASSWORD_VALIDATOR.rules.contains8Chars');
  });

  test('Check whether PasswordValidator Component validates lower and upper case letters when correct data provided', () => {
    const wrapper = shallow(<PasswordValidator value="abcABC" {...minProps} />);
    // Title should be green
    expect(
      wrapper
        .find('PasswordValidator__Row')
        .at(1)
        .prop('style')
    ).toEqual({
      color: '#68c163',
    });
    expect(
      wrapper
        .find('PasswordValidator__Row')
        .at(1)
        .html()
    ).toContain('PASSWORD_VALIDATOR.rules.containsLowerUpperCase');
  });

  test('Check whether PasswordValidator Component validates validates lower and upper case letters when correct data are not provided', () => {
    const wrapper = shallow(
      <PasswordValidator value="1234567" {...minProps} />
    );
    // Title should be grey
    expect(
      wrapper
        .find('PasswordValidator__Row')
        .at(0)
        .prop('style')
    ).toEqual({
      color: '#00000061',
    });
    expect(
      wrapper
        .find('PasswordValidator__Row')
        .at(1)
        .html()
    ).toContain('PASSWORD_VALIDATOR.rules.containsLowerUpperCase');
  });

  test('Check whether PasswordValidator Component validates numbers when correct data provided', () => {
    const wrapper = shallow(<PasswordValidator value="1" {...minProps} />);
    // Title should be green
    expect(
      wrapper
        .find('PasswordValidator__Row')
        .at(2)
        .prop('style')
    ).toEqual({
      color: '#68c163',
    });
    expect(
      wrapper
        .find('PasswordValidator__Row')
        .at(2)
        .html()
    ).toContain('PASSWORD_VALIDATOR.rules.containsNum');
  });

  test('Check whether PasswordValidator Component validates numbers when correct data not provided', () => {
    const wrapper = shallow(<PasswordValidator value="abc" {...minProps} />);
    // Title should be grey
    expect(
      wrapper
        .find('PasswordValidator__Row')
        .at(2)
        .prop('style')
    ).toEqual({
      color: '#00000061',
    });
    expect(
      wrapper
        .find('PasswordValidator__Row')
        .at(2)
        .html()
    ).toContain('PASSWORD_VALIDATOR.rules.containsNum');
  });

  test('Check whether PasswordValidator Component validates no blank space when correct data provided', () => {
    const wrapper = shallow(<PasswordValidator value="1" {...minProps} />);
    // Title should be green
    expect(
      wrapper
        .find('PasswordValidator__Row')
        .at(3)
        .prop('style')
    ).toEqual({
      color: '#68c163',
    });
    expect(
      wrapper
        .find('PasswordValidator__Row')
        .at(3)
        .html()
    ).toContain('PASSWORD_VALIDATOR.rules.containsNoBlank');
  });

  test('Check whether PasswordValidator Component validates no blank space when correct data not provided', () => {
    const wrapper = shallow(
      <PasswordValidator value="Hello World" {...minProps} />
    );
    // Title should be grey
    expect(
      wrapper
        .find('PasswordValidator__Row')
        .at(3)
        .prop('style')
    ).toEqual({
      color: '#00000061',
    });
    expect(
      wrapper
        .find('PasswordValidator__Row')
        .at(3)
        .html()
    ).toContain('PASSWORD_VALIDATOR.rules.containsNoBlank');
  });
});
