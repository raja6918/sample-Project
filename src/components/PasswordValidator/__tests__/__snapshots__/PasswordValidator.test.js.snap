// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test PasswordValidator Component PasswordValidator Component renders correctly 1`] = `
<PasswordValidator__Container>
  <PasswordValidator__Row
    key="1"
    style={
      Object {
        "color": "#00000061",
      }
    }
  >
    <PasswordValidator__DotIcon>
      â€¢
    </PasswordValidator__DotIcon>
    PASSWORD_VALIDATOR.rules.contains8Chars
  </PasswordValidator__Row>
  <PasswordValidator__Row
    key="2"
    style={
      Object {
        "color": "#68c163",
      }
    }
  >
    <Memo([object Object])
      color="inherit"
      fontSize="small"
      style={
        Object {
          "paddingRight": "10px",
          "width": "25px",
        }
      }
    />
    PASSWORD_VALIDATOR.rules.containsLowerUpperCase
  </PasswordValidator__Row>
  <PasswordValidator__Row
    key="3"
    style={
      Object {
        "color": "#68c163",
      }
    }
  >
    <Memo([object Object])
      color="inherit"
      fontSize="small"
      style={
        Object {
          "paddingRight": "10px",
          "width": "25px",
        }
      }
    />
    PASSWORD_VALIDATOR.rules.containsNum
  </PasswordValidator__Row>
  <PasswordValidator__Row
    key="4"
    style={
      Object {
        "color": "#68c163",
      }
    }
  >
    <Memo([object Object])
      color="inherit"
      fontSize="small"
      style={
        Object {
          "paddingRight": "10px",
          "width": "25px",
        }
      }
    />
    PASSWORD_VALIDATOR.rules.containsNoBlank
  </PasswordValidator__Row>
</PasswordValidator__Container>
`;
